torch/utils/data/dataloader.py

```diff
@@ -105,10 +105,14 @@ def _get_distributed_settings():
         return 1, 0
 
 
-def _sharding_worker_init_fn(worker_init_fn, world_size, rank_id, worker_id):
-    global_worker_id = worker_id
+def _sharding_worker_init_fn(is_iter, worker_init_fn, world_size, rank_id, worker_id):
     info = torch.utils.data.get_worker_info()
-    total_workers = info.num_workers
+    if is_iter:
+        global_worker_id = worker_id
+        total_workers = info.num_workers
+    else:
+        global_worker_id = 0
+        total_workers = 1
```

Comment:
We only need to apply distributed + mp sharding for `IterDataPipe`. For `MapDataPipe`, it seems distributed sharding is the only requirement.

Reply:
It seems like it is possible that users start off with `IterDataPIpe`, add a `sharding_filer`, then convert to `MapDataPipe` prior to passing it into `DataLoader`.

IIUC this might not handle that case properly? We can potentially prevent that case with a linter?

Reply:
It should be fine. And, it's a great idea to add a test to guard the behavior. Let me do that

Reply:
Will sharding still be correctly applied if the last DataPipe that user provides is `to_map_datapipe`?

Reply:
I think so because `DataLoader` would shard the indices in the main process. So, as long as the end DataPipe is a `MapDataPipe`, the whole pipeline should be sharded based on number of workers.

I might not be able to add a test here due to `to_map_datapipe` is in TorchData. But, I will definitely verify the behavior locally.

Reply:
It turns out `shuffle` is way more complicated than I though. I might need to fully revamp the logic of this PR. Need to do more investigation!

Reply:
BTW, it's also blocked by this issue: https://github.com/pytorch/data/issues/753